python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+
+_score(covariance,X,weights=None)
+
+_score_samples(covariance,X)
+
+----------------------------
+
+_entropy(covariance)
+
+_entropy_subset(covariance,idxs)
+
+_mutual_information(covariance,idxs1,idxs2):
+
+"""
+import numpy as np
+from numpy import linalg as la
+from datetime import datetime as dt
+import pytz
+
+class BaseCovariance(object):
+    """
+    This object basically contains all the methods asides for fitting
+    a covariance matrix.
+    """
+
+    def __init__(self):
+        self.creationDate = dt.now(pytz.timezone("US/Pacific"))
+
+    #################
+    # Miscellaneous #
+    #################
+    def get_precision(self):
+        return _get_precision(self.covariance)
+
+    def get_stable_rank(self):
+        return _get_stable_rank(self.covariance)
+
+    def predict(self, Xobs, idxs):
+        return _predict(self.covariance, Xobs, idxs)
+
+    #####################################
+    # Gaussian Likelihood-based methods #
+    #####################################
+    def conditional_score(self, X, idxs, weights=None):
+        return _conditional_score(self.covariance, X, idxs, weights=weights)
+
+    def conditional_score_samples(self, X, idxs):
+        return _conditional_score_samples(self.covariance, X, idxs)
+
+    def marginal_score(self, X, idxs, weights=None):
+        return _marginal_score(self.covariance, X, idxs, weights=weights)
+
+    def marginal_score_samples(self, X, idxs):
+        return _marginal_score_samples(self.covariance, X, idxs)
+
+    def score(self, X, weights=None):
+        return _score(self.covariance, X, weights=weights)
+
+    def score_samples(self, X):
+        return _score_samples(self.covariance, X)
+
+    #################################
+    # Information-theoretic methods #
+    #################################
+    def entropy(self):
+        return _entropy(self.covariance)
+
+    def entropy_subset(self, idxs):
+        return _entropy_subset(self.covariance, idxs)
+
+    def mutual_information(self, idxs1, idxs2):
+        return _mutual_information(self.covariance, idxs1, idxs2)
+
+
+###########################################
+###########################################
+###########################################
+###                                     ###
+###                                     ###
+###   Methods for covariance matrices   ###
+###                                     ###
+###                                     ###
+###########################################
+###########################################
+###########################################
+
+
+def _get_precision(covariance):
+    """
+    Gets the precision matrix defined as the inverse of the covariance
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Returns
+    -------
+    precision : np.array-like(sum(p),sum(p))
+        The inverse of the covariance matrix
+    """
+    precision = la.inv(covariance)
```

Comment:
I can tell from the documentation above that you've already thought about this, but are we sure we need to invert a (p, p) matrix?  What are typical values of p in this case?  Depending on what we're doing with the covariance matrix downstream, we may be able to get away with doing something else that's more numerically stable.

Not a blocking comment rn, just curious

Reply:
Unfortunately in statistics often times the explicit estimates of the precision matrix are of interest for interpretation. Certainly though if you want to evaluate the likelihood this is a bad idea.